encoding/codecv0.go

```
@@ -0,0 +1,409 @@
+
+	totalL1MessagePoppedBefore := b.TotalL1MessagePoppedBefore
+
+	for _, chunk := range b.Chunks {
+		chunkL1CommitGas, err := o.EstimateChunkL1CommitGas(chunk)
+		if err != nil {
+			return 0, err
+		}
+		totalL1CommitGas += chunkL1CommitGas
+
+		totalL1MessagePoppedInChunk := chunk.NumL1Messages(totalL1MessagePoppedBefore)
+		totalL1MessagePoppedBefore += totalL1MessagePoppedInChunk
+
+		totalL1CommitGas += CalldataNonZeroByteGas * (32 * (totalL1MessagePoppedInChunk + 255) / 256)
+		totalL1CommitGas += GetKeccak256Gas(89 + 32*(totalL1MessagePoppedInChunk+255)/256)
+
+		chunkL1CommitCalldataSize, err := o.EstimateChunkL1CommitCalldataSize(chunk)
+		if err != nil {
+			return 0, err
+		}
+		totalL1CommitGas += GetMemoryExpansionCost(chunkL1CommitCalldataSize)
+	}
+
+	return totalL1CommitGas, nil
+}
+
+// EstimateBatchL1CommitCalldataSize calculates the calldata size in l1 commit for this batch approximately.
+func (o *DACodecV0) EstimateBatchL1CommitCalldataSize(b *Batch) (uint64, error) {
+	var totalL1CommitCalldataSize uint64
+	for _, chunk := range b.Chunks {
+		chunkL1CommitCalldataSize, err := o.EstimateChunkL1CommitCalldataSize(chunk)
+		if err != nil {
+			return 0, err
+		}
+		totalL1CommitCalldataSize += chunkL1CommitCalldataSize
+	}
+	return totalL1CommitCalldataSize, nil
+}
+
+// CheckChunkCompressedDataCompatibility checks the compressed data compatibility for a batch built from a single chunk.
+func (o *DACodecV0) CheckChunkCompressedDataCompatibility(c *Chunk) (bool, error) {
+	return true, nil
+}
+
+// CheckBatchCompressedDataCompatibility checks the compressed data compatibility for a batch.
+func (o *DACodecV0) CheckBatchCompressedDataCompatibility(b *Batch) (bool, error) {
+	return true, nil
+}
+
+// EstimateChunkL1CommitBatchSizeAndBlobSize estimates the L1 commit uncompressed batch size and compressed blob size for a single chunk.
+func (o *DACodecV0) EstimateChunkL1CommitBatchSizeAndBlobSize(c *Chunk) (uint64, uint64, error) {
```

I'm wondering whether it would be better to return an error for these methods that are not supported on a codec. If this happens this might be even a coding error and thus even a panic might be justified here as it would greatly help when developing and calling the wrong things on a codec.

---

it will simplify the usage in https://github.com/scroll-tech/scroll/pull/1530. otherwise will add many if conditions. we can discuss this further.